### PR TITLE
[Operator] Optimize CrossEntropyLoss

### PR DESCRIPTION
**Pull request by StrongSpoon**
1. reimplement cross_entropy_loss forward and backward
2. support indices/probabilities/weight/reduction/ignore_index/label_smoothing
3. perform better than torch eager on large scale tensors
截屏2024-07-24 10 50 41

### CODE REVIEW

File: src/flag_gems/ops/celoss.py [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+        out = torch.empty(shape, dtype=inp.dtype, device=inp.device)
+        grid = lambda meta: (N, triton.cdiv(D, meta["BLOCK_D"]))
+
+        if target.ndim == dim:
+            # target probabilities
+            with torch.cuda.device(inp.device):
+                celoss_probability_kernel[grid](inp, tgt, weight, out, N, C, D)
+        else:
+            # target indices
+            w_tgt = torch.empty(shape, dtype=inp.dtype, device=inp.device)
```
Comment (iclementine) [inline src/flag_gems/ops/celoss.py]: Is this (`w_tgt`) only for the purpose of debugging?

Comment (StrongSpoon) [reply src/flag_gems/ops/celoss.py]: reduction mode 'mean' needs it

**Review by tongxin — DISMISSED** [review]
LG. Are you going to add label smoothing later?

**Comment by StrongSpoon** [comment]
> LG. Are you going to add label smoothing later?

Fisinehd

**Review by Bowen12992 — APPROVED** [review]
LGTM
